core/src/main/java/com/linecorp/armeria/server/VirtualHostBuilder.java

```diff
@@ -917,17 +911,24 @@ VirtualHost build(VirtualHostBuilder template) {
         assert accessLoggerMapper != null;
         assert extensions != null;
 
-        virtualHostAnnotatedServiceBindingBuilders.stream()
-                                                  .flatMap(b -> b.buildServiceConfigBuilder(extensions)
-                                                                 .stream())
-                                                  .forEach(this::addServiceConfigBuilder);
-
-        final List<ServiceConfigBuilder> serviceConfigBuilders =
-                getServiceConfigBuilders(template);
-        final List<ServiceConfig> serviceConfigs = serviceConfigBuilders.stream().map(cfgBuilder -> {
-            return cfgBuilder.build(requestTimeoutMillis, maxRequestLength, verboseResponses,
-                                    accessLogWriter, shutdownAccessLogWriterOnStop);
-        }).collect(toImmutableList());
+        final List<ServiceConfig> serviceConfigs = getServiceConfigSetters(template)
+                .stream()
+                .flatMap(cfgSetters -> {
+                    if (cfgSetters instanceof VirtualHostAnnotatedServiceBindingBuilder) {
+                        return ((VirtualHostAnnotatedServiceBindingBuilder) cfgSetters)
+                                .buildServiceConfigBuilder(extensions).stream();
+                    } else if (cfgSetters instanceof AnnotatedServiceBindingBuilder) {
+                        return ((AnnotatedServiceBindingBuilder) cfgSetters)
+                                .buildServiceConfigBuilder(extensions).stream();
+                    } else if (cfgSetters instanceof ServiceConfigBuilder) {
+                        return Stream.of((ServiceConfigBuilder) cfgSetters);
+                    } else {
+                        throw new Error(); // Should not reach here.
```

Comment:
Error is too generic in here may be we should throw more specific Exception in here

Reply:
This is an internal API so users must not encounter this error. But, let me improve the error message.